### ISSUE TITLE
docker client issue when setting Firewall rule to "tcp PORT all"

### ISSUE THREAD

**Issue by moogle19** [joyent/sdc-docker#73]
If you create a docker container and add a Firewall rule that allows traffic from "any" to the docker vm on "tcp PORT all" the docker container is reachable from the outside, but the docker ps command fails with 
```
json: cannot unmarshal string into Go value of type int
```

Firewall rule:
<img width="538" alt="screen shot 2016-03-22 at 3 22 49 pm" src="https://cloud.githubusercontent.com/assets/792046/13954658/0a9c3334-f042-11e5-8344-a594952ca55f.png">

That is because the values in the ports array for PrivatePort and PublicPort are `"any"`, but the docker client expects an integer.

curl output for `/containers/json?all=1`
```
[{
	"Id": "88993aa89a3b4de7b20617babc274b73ecb593bd914340e0a9618c78037b565b",
	"Created": 1458654290,
	"Command": "/assets/wrapper",
	"Names": ["/compassionate_perlman"],
	"Status": "Exited (127) 4 minutes ago",
	"Image": "gitlab/gitlab-ce",
	"Labels": {
		"com.joyent.package": "sample-1G"
	},
	"Ports": [{
		"IP": "0.0.0.0",
		"PrivatePort": "all", <-- here
		"PublicPort": "all",  <-- here
		"Type": "tcp"
	}, {
		"PrivatePort": 22,
		"Type": "tcp"
	}, {
		"PrivatePort": 443,
		"Type": "tcp"
	}, {
		"PrivatePort": 80,
		"Type": "tcp"
	}]
}]
```

**Comment by notmatt** [comment]
@moogle19 Thanks! I've opened an internal bug on this [here](https://smartos.org/bugview/DOCKER-760).

The root cause of your specific problem is that when we map firewall rules to the docker port list, we aren't ensuring that we're displaying only `int`s for ports, which is what's required by the [docker Port type](https://github.com/docker/engine-api/blob/9524d7ae81ff55771852b6269f40f2a878315de9/types/types.go#L137-L142).

Unfortunately, this also means that there aren't any extremely pleasing solutions to the problem of what to do for `all`. A list of 65k published ports (what docker does for port ranges) would likely be problematic for everyone (at least if you don't expect it). An int outside the normal port range (like '-1') that had special semantics in SDC would work on the SDC end, but making sure it didn't break anything in the docker client would be quite difficult (not to mention creating a portability problem).

So the best idea we have at the moment is probably not to translate firewall rules that don't have good representations in the docker container format. You would still be able to apply them via the SDC UI or APIs, but they wouldn't be reported by the docker client. Would that work for your usecase?

thanks,
matt.

**Comment by magnayn** [comment]
I don't know about the OP, but that would work for me. At the moment any client listing containers dies because it can't parse "all" as an int, so it'd be better for that to work at the expense of not neccesarily knowing which ports were open.

**Comment by melloc** [comment]
This is fixed in 3e42224 which is in this week's release.
